clis/teliod/src/nc.rs

```diff
@@ -0,0 +1,533 @@
+    TokenHttpError(#[from] reqwest::Error),
+    #[error("Failed to get token: {0}")]
+    TokenError(String),
+    #[error("Malformed notification center json response: {0}")]
+    MalformedNotificationCenterJsonResponse(#[from] serde_json::Error),
+    #[error("Mqtt broker rejected connection: {0:?}")]
+    MqttConnectionRejected(ConnectReturnCode),
+    #[error("Unexpected mqtt event before connection established: {0:?}")]
+    MqttUnexpectedEventBeforeConnection(Event),
+    #[error("Mqtt connection timeout")]
+    MqttConnectionTimeout,
+    #[error("Mqtt connection error: {0}")]
+    MqttConnectionError(rumqttc::ConnectionError),
+    #[error(transparent)]
+    BackoffBoundsError(#[from] BackoffError),
+    #[error(transparent)]
+    MqttClientError(#[from] ClientError),
+    #[error("Notification Center credentials are missing hostname")]
+    CredentialsMissingHostname,
+    #[error("Notification Center credentials are missing port number")]
+    CredentialsMissingPortNumber,
+    #[error("Failed to read http cert file: {0}")]
+    FailedToReadHttpCertFile(std::io::Error),
+    #[error("Failed to read mqtt cert file: {0}")]
+    FailedToReadMqttCertFile(std::io::Error),
+    #[error("Failed to parse mqtt pem certificate: {0:?}")]
+    FailedToParseMqttCertificate(tokio_rustls::rustls::pki_types::pem::Error),
+    #[error("Failed to load native certs: {0:?}")]
+    FailedToLoadNativeCerts(Vec<rustls_native_certs::Error>),
+}
+
+type Callback = Arc<dyn Fn(&[Uuid]) + Send + Sync>;
+
+pub struct NotificationCenter {
+    callbacks: Arc<Mutex<Vec<Callback>>>,
+}
+
+struct NCConfig {
+    authentication_token: String,
+    app_user_uid: Uuid,
+    callbacks: Arc<Mutex<Vec<Callback>>>,
+    http_certificate_file_path: Option<PathBuf>,
+    mqtt_certificate_file_path: Option<PathBuf>,
+}
+
+impl NotificationCenter {
+    pub async fn new(
+        config: &super::TeliodDaemonConfig,
+        callbacks: Vec<Callback>,
+    ) -> Result<Self, Error> {
+        let callbacks = Arc::new(Mutex::new(callbacks));
+
+        let nc_config = NCConfig {
+            authentication_token: config.authentication_token.clone(),
+            app_user_uid: config.app_user_uid,
+            callbacks: callbacks.clone(),
+
+            http_certificate_file_path: config.http_certificate_file_path.clone(),
+            mqtt_certificate_file_path: config.mqtt_certificate_file_path.clone(),
+        };
+
+        start_mqtt(nc_config).await?;
+
+        Ok(Self { callbacks })
+    }
+
+    pub async fn add_callback(&self, callback: Callback) {
+        let mut callbacks = self.callbacks.lock().await;
+        callbacks.push(callback);
+    }
+}
+
+async fn start_mqtt(nc_config: NCConfig) -> Result<(), Error> {
+    let backoff_bounds = ExponentialBackoffBounds {
+        initial: Duration::from_secs(1),
+        maximal: Some(Duration::from_secs(300)),
+    };
+    let backoff = ExponentialBackoff::new(backoff_bounds)?;
```

Comment:
This would mean we construct new backoff multiple times **and** have to deal with the fact that this construction can fail in the async task. I think that keeping it here makes error handling much simpler than having to deal with error propagation from the task.